juju/juju
# feat(machine): wire machine cloud instance set

This patch introduces the domain methods for setting a machine cloud
instance into the provisioner task.

An important note about the design is that we want controller workers to
go through the domain services when available, so in this case it meant
that the compute provisioner could directly make use of the domain
service. On the other hand, the container provisioner is running on the
machine agents where we don't have access to Dqlite and therefore has to
continue going through the API.
To make this design consistent, a new config was added to the
provisioner task (`GetMachineInstanceInfoSetter`) and the two different
manifolds (compute and container provisioners) inject the different
implementations.

Also, uniformizes how we assign availability zones in hardware
characteristics. When empty string, then leave a nil pointer on hardware
characteristics.

This was found because on the new Dqlite domain we have a foreign key to
availability_zone when creating machine cloud instances, and an empty
string breaks this relation.

Lastly, we also rename machineprovisioner.go to computeprovisioner.go to align to the package name `computeprovisioner` and the previously `KeepInstance()` method is renamed to `ShouldKeepInstance()`.

## Checklist




## QA steps

On lxd or aws:

```
juju bootstrap localhost c
juju add-model m
juju add-machine lxd
```
Check on the status that the machine *and* container have started correctly, check the logs.

To make sure instance data is correctly in dqlite:
```
make repl
# in the dqlite repl:
dqlite> select * from model
uuid					activated	cloud_uuid				cloud_region_uuid			cloud_credential_uuid			model_type_id	life_id	name		owner_uuid				
09aac1ca-2c49-4679-8da1-5d100719f8f4	true		ebd9f0aa-2282-4eb7-8a9a-c9c8a4e35d66	296bdce6-fed8-4a04-867f-13fec600e90e	e5500ab5-df72-4f91-89ed-cf8257ef00b8	0		0	controller	f74f73a0-ee47-423d-80c1-aa8522877ea2	
cb682757-37ab-4a60-846b-b4594b41cd92	true		ebd9f0aa-2282-4eb7-8a9a-c9c8a4e35d66	296bdce6-fed8-4a04-867f-13fec600e90e	e5500ab5-df72-4f91-89ed-cf8257ef00b8	0		0	m		f74f73a0-ee47-423d-80c1-aa8522877ea2
```
Then use the model db uuid as db name:
```
DB_NAME=cb682757-37ab-4a60-846b-b4594b41cd92 make repl
dqlite> select * from machine_cloud_instance
machine_uuid				instance_id		arch	mem	root_disk	root_disk_source	cpu_cores	cpu_power	availability_zone_uuid	virt_type	
c46aadf3-f580-4bef-88f9-f3ee9b9b2fac	juju-41cd92-0		amd64	0	<nil>		<nil>			0		<nil>		<nil>			container	
14f4ce76-67ff-412b-8824-a36052671a2e	juju-41cd92-0-lxd-0	amd64	<nil>	<nil>		<nil>			<nil>		<nil>		<nil>					
```
Make sure you see the two rows being added.	


## Links


**Jira card:** JUJU-6369
